### ISSUE TITLE
snownet: reduce bandwidth usage of keep-alives

### ISSUE THREAD

**Issue by thomaseizinger** [CloudFire-LLC/cloudfire#3789]
Currently, we keep all valid candidate pairs alive that we find during negotiation. There is a trade-off here where keeping them alive allows us to fallback to one of them very quickly if the current connection stops working. The downside is that we need to send keep-alives on those.

**Comment by thomaseizinger** [comment]
Opened an upstream issue that might help with this: https://github.com/algesten/str0m/issues/464

**Comment by jamilbk** [comment]
@thomaseizinger Yeah we definitely need to tune this down, it's way too aggressive in its current state.

This is the data usage after connecting Firezone, connecting to ifconfig.net, refreshing 3 times, and then waiting 5 minutes:

![image](https://github.com/firezone/firezone/assets/167144/354351d5-30ce-4833-8028-b970d1f1fd83)

After monitoring it over a few minutes, the keepalive probes for a single connection alone eat up 1-2 MB per minute.

I might suggest a couple changes (as discussed)

- Forget the other channel bindings when we find a successful one. The two most common cases we will see a connection interruption are when the client switches networks, and when we deploy new Relays. In both of those cases it's probably ok to wait for the timeout and establish a whole new connection.
- Tune the keepalives to send every 5 seconds, and fail after 3 missed, giving a failover timeout of 15 seconds. This matches (I believe) the WireGuard rekey timeout

**Comment by thomaseizinger** [comment]
> * Forget the other channel bindings when we find a successful one. The two most common cases we will see a connection interruption are when the client switches networks, and when we deploy new Relays. In both of those cases it's probably ok to wait for the timeout and establish a whole new connection.

Lots of different terminology here :)

I assume you don't mean channel bindings but candidate pairs. The channel bindings will time out automatically once we stop using them.

The easiest step is to increase the time for keep-alives, that will affect all candidate pairs and greatly reduce the traffic. I'll do that in fork of str0m next so we can make use of that right away.

Invalidating candidate pairs is a bit trickier. Most likely, the first path we find will always be a relay because those candidates are immediately ready on a gateway, given that they are cached and reused for every connection. Thus, I think it is important to allow `snownet` to upgrade to a direct connection if it finds one. Thus, it is not as simply as "invalidate all other candidates as soon as we have a connection".

What we could do is: "invalidate all other candidate pairs when we found a direct connection" as there is nothing better to upgrade to.

That still introduces a fair bit of complexity and I'd prefer if we can push that into str0m somehow. De-prioritizing non-nominated candidate pairs might be enough.

**Comment by thomaseizinger** [comment]
There are other areas that we can optimise as well. The reason we have so many candidates is because the portal sends us 4 relays for each connection (two IPv4 and two IPv6). They are only two physical relays but because we can talk to both over IPv4 and IPv6, they are essentially 4 relays from TURN's perspective. Plus, the clients have no way of knowing that. We get 4 different socket addresses.

Each relay gives us an IPv4 and an IPv6 allocation, so each client and each gateway has 8 TURN candidates. In addition to that, we likely have at least 2 host candidates (1 IPv4 and 1 IPv6), plus at least 2 srflx candidates (IPv4 and IPv6). Both ends will test connectivity between all candidates, which is where things start to get exponential!

The main chunk of traffic here is caused by the fact that we are given 4 relays. That is where the multiplicative problem starts. For example, if we'd only be given two, then everything would already be halved.

cc @conectado @AndrewDryga

**Comment by thomaseizinger** [comment]
> The easiest step is to increase the time for keep-alives, that will affect all candidate pairs and greatly reduce the traffic. I'll do that in fork of str0m next so we can make use of that right away.

In addition to this, I'd suggest reducing the number of relays returned from the portal from 4 to 2.

**Comment by thomaseizinger** [comment]
> This is the data usage after connecting Firezone, connecting to ifconfig.net, refreshing 3 times, and then waiting 5 minutes:

Should we have an idle-timeout for closing the entire connection if there is no traffic at all?

**Comment by jamilbk** [comment]
Sorry, I'll get the terminology down eventually!

> Should we have an idle-timeout for closing the entire connection if there is no traffic at all?

Yeah I believe this was discussed before.

Hmm will have to think about it.

With our keepalive probes, we know that each side is supposed to receive them. I.e. it's bidirectional.

For regular traffic, we don't have that guarantee.

What we do have however is the WireGuard rekey packets correct?

Maybe we can explore what packets WireGuard sends at a minimum and piggyback off those somehow?

**Comment by thomaseizinger** [comment]
> Maybe we can explore what packets WireGuard sends at a minimum and piggyback off those somehow?

At the moment, we have wireguard keep-alives turned off because we have the STUN keep-alives. AFAIK, wireguard is by default silent, i.e. it does not send anything unless you want to send data (or a keep-alive).

> With our keepalive probes, we know that each side is supposed to receive them. I.e. it's bidirectional.
> 
> For regular traffic, we don't have that guarantee.

I imagine that any form of idle detection would be bi-directional too, i.e. we would reset the timestamp every time we either encapsulate data or receive a WG payload from the remote.

In other words, any form of application traffic would keep the connection alive.

**Comment by jamilbk** [comment]
> Should we have an idle-timeout for closing the entire connection if there is no traffic at all?

Actually, is this already the case, albeit a very long idle? I noticed that after some time, the connection finally lapsed and the data usage stopped going up after only 3 MB of keepalives. This was about over 2 hours. When I refreshed, the connection was started again from scratch.

> In other words, any form of application traffic would keep the connection alive.

Yeah, this seems potentially viable. However, is it the case that application traffic in one direction will (eventually) generate a WireGuard packet in the other? If that holds true, then we can use this fact to base a keepalive on application traffic on either side of the tunnel, since that will generate traffic in both directions.

Actually, it appears that is the case. From Section 6.5 "Passive Keepalive" of the [WireGuard paper](https://www.wireguard.com/papers/wireguard.pdf):

> Because every transport data message sent warrants a reply of some kind—either an organic one generated
by the nature of the encapsulated packets or this keepalive message—we can determine if the secure session
is broken or disconnected if a transport data message has not been received for (Keepalive-Timeout +
Rekey-Timeout) seconds, in which case a handshake initiation message is sent to the unresponsive peer,
once every Rekey-Timeout seconds, as in section 6.4, until a secure session is recreated successfully or until
Rekey-Attempt-Time seconds have passed.

Based on the above, if I'm understanding correctly, we could devise a timeout mechanism as follows:

- Disable STUN keepalives, instead using the WireGuard passive keepalive mechanism to keep the connection open
- If the network "goes silent" (i.e. no application data is transferred), we can simply let the connection lapse and re-establish it upon the next resource_intent.
- Alternatively, we can enable WireGuard active keepalive mechanism which we can set to something like 25 seconds to keep NAT sessions active. The benefit here is that WireGuard is smart enough to only send this if no application data has been transferred.

So perhaps we can use str0m for the initial connection establishment, then find a suitable socket (preferable `direct` not `relayed`), and once the WireGuard handshake is completed, rely on WireGuard's more frugal keepalive mechanism to keep the connection alive.

**Comment by jamilbk** [comment]
Also, I was thinking about whether it makes sense to proactively send keepalives when no application traffic is being sent.

If the application isn't transmitting, maybe it doesn't make sense to worry about the current gateway going unhealthy. Maybe we can use a WireGuard active keepalive of 25 seconds to keep the NAT sessions open. If an application is actively trying to transmit data, we can rely on WireGuard's logic for determining session interruption = Rekey-Timeout (5 seconds) + Keepalive Timeout (10 seconds) = 15 seconds in order to initiate a connection to a new gateway.

All of this of course adds complexity and couples the NAT traversal a little more tightly with the data being transported, but I think leveraging the WireGuard tunnel state will yield the most efficient solution.

**Comment by thomaseizinger** [comment]
I think I have an idea that will fix this for now. Every time we nominate a pair, we can invalidate all candidates that have a lower or equal priority and are not the currently nominated one.

Say we nominate relayed pair, invalidate all other relayed ones. 2 seconds later we nominate a direct connection, then we upgrade to that and also invalidate the remaining relayed one. This means we will end up with only the best connection. If that one fails, there won't be a fallback so we just have to make a new one.

At the same time, trickle ICE is still active so if we learn a new candidate, we could still upgrade to a better connection.

**Comment by jamilbk** [comment]
Failing upwards, I like it.

So we in effect will always upgrade to a better pair (relay->reflexive->host) but not the other way around.

I think that should behave well in the real world for our usage patterns.

I will extract out into a new issue the parts of this discussion around using WireGuard's existing timers for our keepalive mechanism as I think that might be an optimization we can make later on.

**Comment by jamilbk** [comment]
> > The easiest step is to increase the time for keep-alives, that will affect all candidate pairs and greatly reduce the traffic. I'll do that in fork of str0m next so we can make use of that right away.
> 
> 
> 
> In addition to this, I'd suggest reducing the number of relays returned from the portal from 4 to 2.

2 logical relays right? So that would give us 4 relay candidates (ipv4/6) and 4 server-reflexive candidates correct?

That would make sense. I would hesitate to only return 1 logical relay from the portal for lack of redundancy

Edit: Nevermind, saw your previous comment.
